lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp

```diff
@@ -223,28 +223,39 @@ bool LoopPipelinerInternal::verifySchedule() {
   // Pre-compute the unrolled cycle of each op.
   DenseMap<Operation *, int64_t> unrolledCyles;
   for (int64_t cycle = 0; cycle < numCylesPerIter; cycle++) {
-    Operation *def = opOrder[cycle];
-    auto it = stages.find(def);
+    Operation *op = opOrder[cycle];
+    auto it = stages.find(op);
     assert(it != stages.end());
     int64_t stage = it->second;
-    unrolledCyles[def] = cycle + stage * numCylesPerIter;
+    op->walk<WalkOrder::PreOrder>([&](Operation *def) {
+      unrolledCyles[def] = cycle + stage * numCylesPerIter;
+    });
   }
-  for (Operation *consumer : opOrder) {
-    int64_t consumerCycle = unrolledCyles[consumer];
-    for (Value operand : consumer->getOperands()) {
-      auto [producer, distance] = getDefiningOpAndDistance(operand);
-      if (!producer)
-        continue;
-      auto it = unrolledCyles.find(producer);
-      // Skip producer coming from outside the loop.
-      if (it == unrolledCyles.end())
-        continue;
-      int64_t producerCycle = it->second;
-      if (consumerCycle < producerCycle - numCylesPerIter * distance) {
-        consumer->emitError("operation scheduled before its operands");
-        return false;
+
+  for (Operation *op : opOrder) {
+    bool valid = false;
+    op->walk<WalkOrder::PreOrder>([&](Operation *consumer) {
```

Comment:
That's a bit of a nit but can we just collect all the operands that are from `ForOp` op and replace that instead of using `consumer->getOperands()` then the rest of the logic can stay the same? That might be less confusing as we don't really schedule nested ops so assigning a cycle to them is a bit weird.

Basically something like:
```
SmallVector<Value> getNestedLoopOperands(Operation* op) {
  SmallVector<Value> operands;
  op->walk([&](Operation* nestedOp) {
    for (Value operand : nestedOp->getOperands()) {
      if (operand.getParentBlock() == op->getBlock())
        operands.push_back(operand);
    }
  }
}
....
for (Operation *consumer : opOrder) {
    int64_t consumerCycle = unrolledCyles[consumer];
    for (Value operand : getNestedLoopOperands(consumer))
        ...
```

Reply:
Makes a lot of sense, I like it! Let me try that out.

Reply:
Updated! I have changed the code to catch the operands coming from any nesting levels, not just where the `op` was defined. I think this should be correct, let me know if you think otherwise.

Reply:
yeah I guess that is filtered later. You can even skip the `if (operand.getParentBlock()->getParentOp()->isAncestor(nestedOp))` that will avoid having to recurse through the stack of regions

Reply:
Actually, I think I'll make an Utility function out of this one, I see more places where I'll need to use it for similar reasons.
On top of this, I think we should actually return SetVector instead of Vector, right? We don't want the operands to be duplicated.

Reply:
Ouch, do we want to use an utility function in expander though...? That would make it harder to upstream it...

Reply:
yes let's not add dependency outside of this file. You can make a utility inside the file if you think it helps. I wouldn't overthink it ;)

Reply:
I don't think duplicates are a problem actually. I would have left it as SmallVector